### DISCUSSION TITLE
How can server implementation stub determine the IP address of a client

### DISCUSSION THREAD

**Discussion by srl100** [COVESA/vsomeip#712] [Q&A] [answered]
As detailed in issue [https://github.com/COVESA/vsomeip/issues/679](https://github.com/COVESA/vsomeip/issues/679), in a vsomeip/CommonAPI stub implementation, how can the server determine the IP address of the client that triggers a call to a method?

**Comment by Man777** [comment]
Did you already got running between 2 different clients, so that you can observe with wireshark the trace between them ? Project with fidl and generated with cgen and build etc.

**Reply by srl100** [reply]
Yes, I already have a vsomeip/CommonAPI server running on an embedded Linux target, and can exercise it from a client running under Windows. I can use the client to request FIDL-specified methods implemented on the server and the requests and responses over the wire are as expected.

However, I need to be able to determine the IP address of the client within my server implementation but it appears that the _client->getHostAddress() method within the stub returns "0.0.0.0" rather than the IP address of the client.

The signature of my stub function is:

`void VideoCameraStubImpl::setCamExclusive(const std::shared_ptr<CommonAPI::ClientId> _client, uint32_t _ttl, setCamExclusiveReply_t _reply)`

Digging into the CommonAPI Types.hpp file, ClientId is documented as:

```
/**
 * \brief Identifies a client sending a call to a stub.
 *
 * The ClientId is used to identify the caller within a stub.
 * The ClientId is supposed to be added by the middleware and can be compared using the == operator.
 */
```
...so something in the middleware is not filling in the ClientId correctly.

I've added a breakpoint to line 42 of CommonAPI/SomeIP/ClientID.cpp, and it appears that the constructor is getting called with _sec_client->host = 0, so I suspect that something further up the call stack is not determining the source address of the UDP message.

The backtrace is:

```
#0  CommonAPI::SomeIP::ClientId::ClientId (this=0xffffe8000d90,
    _client=<optimized out>, _sec_client=0xfffff56ede00, _env=...)
    at /home/user/capicxx-someip-runtime/src/CommonAPI/SomeIP/ClientId.cpp:42
#1  0x0000aaaaaaad3c4c in std::_Construct<CommonAPI::SomeIP::ClientId, unsigned short, vsomeip_sec_client_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (__p=0xffffe8000d90)
    at /usr/include/c++/12.3.0/bits/stl_construct.h:119
#2  std::allocator_traits<std::allocator<void> >::construct<CommonAPI::SomeIP::ClientId, unsigned short, vsomeip_sec_client_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (__p=0xffffe8000d90)
    at /usr/include/c++/12.3.0/bits/alloc_traits.h:635
#3  std::_Sp_counted_ptr_inplace<CommonAPI::SomeIP::ClientId, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<unsigned short, vsomeip_sec_client_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (__a=..., this=0xffffe8000d80)
    at /usr/include/c++/12.3.0/bits/shared_ptr_base.h:604
#4  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<CommonAPI::SomeIP::ClientId, std::allocator<void>, unsigned short, vsomeip_sec_client_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >
    (__a=..., __p=<synthetic pointer>: <optimized out>,
    this=<synthetic pointer>)
    at /usr/include/c++/12.3.0/bits/shared_ptr_base.h:971
#5  std::__shared_ptr<CommonAPI::SomeIP::ClientId, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<void>, unsigned short, vsomeip_sec_client_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (
    __tag=..., this=<synthetic pointer>)
    at /usr/include/c++/12.3.0/bits/shared_ptr_base.h:1712
#6  std::shared_ptr<CommonAPI::SomeIP::ClientId>::shared_ptr<std::allocator<void>, unsigned short, vsomeip_sec_client_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (__tag=...,
    this=<synthetic pointer>)
    at /usr/include/c++/12.3.0/bits/shared_ptr.h:464
#7  std::make_shared<CommonAPI::SomeIP::ClientId, unsigned short, vsomeip_sec_client_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > ()
    at /usr/include/c++/12.3.0/bits/shared_ptr.h:1010
#8  CommonAPI::SomeIP::MethodWithReplyStubDispatcher<v1::commonapi::examples::VideoCameraStub, std::tuple<unsigned int>, std::tuple<>, std::tuple<CommonAPI::Som--Type <RET> for more, q to quit, c to continue without paging--
eIP::IntegerDeployment<unsigned int> >, std::tuple<> >::dispatchMessageHelper<0ul>(CommonAPI::SomeIP::Message const&, std::shared_ptr<v1::commonapi::examples::VideoCameraStub> const&, v1::commonapi::examples::VideoCameraStubRemoteEvent*, std::shared_ptr<CommonAPI::SomeIP::ProxyConnection>, CommonAPI::SomeIP::index_sequence<0ul>, CommonAPI::SomeIP::index_sequence<>) (
    this=this@entry=0xaaaaaab1e1f0, _message=..., _stub=...,
    _remoteEventHandler=_remoteEventHandler@entry=0xaaaaaab0e548,
    _connection=...)
    at /home/user/capicxx-someip-runtime/include/CommonAPI/SomeIP/StubAdapterHelper.hpp:464
#9  0x0000aaaaaaad4020 in CommonAPI::SomeIP::MethodWithReplyStubDispatcher<v1::commonapi::examples::VideoCameraStub, std::tuple<unsigned int>, std::tuple<>, std::tuple<CommonAPI::SomeIP::IntegerDeployment<unsigned int> >, std::tuple<> >::dispatchMessage(CommonAPI::SomeIP::Message const&, std::shared_ptr<v1::commonapi::examples::VideoCameraStub> const&, v1::commonapi::examples::VideoCameraStubRemoteEvent*, std::shared_ptr<CommonAPI::SomeIP::ProxyConnection>) (
    this=0xaaaaaab1e1f0, _message=..., _stub=...,
    _remoteEventHandler=0xaaaaaab0e548, _connection=...)
    at /home/user/capicxx-someip-runtime/include/CommonAPI/SomeIP/StubAdapterHelper.hpp:396
#10 0x0000aaaaaaac6e58 in CommonAPI::SomeIP::StubAdapterHelper<v1::commonapi::examples::VideoCameraStub>::findDispatcherAndHandle (this=0xaaaaaab1dfb0,
    _message=..., _method=<optimized out>)
    at /home/user/capicxx-someip-runtime/include/CommonAPI/SomeIP/StubAdapterHelper.hpp:148
#11 0x0000aaaaaaac705c in virtual thunk to CommonAPI::SomeIP::StubAdapterHelper<v1::commonapi::examples::VideoCameraStub>::onInterfaceMessage(CommonAPI::SomeIP::Message const&) ()
    at /home/user/capicxx-someip-runtime/include/CommonAPI/SomeIP/StubAdapterHelper.hpp:135
#12 0x0000fffff7f25820 in CommonAPI::SomeIP::StubManager::handleMessage (
    this=0xaaaaaab0b240, _message=...)
    at /home/user/capicxx-someip-runtime/src/CommonAPI/SomeIP/StubManager.cpp:82
#13 0x0000fffff7eeca50 in std::function<bool (CommonAPI::SomeIP::Message const&)>::operator()(CommonAPI::SomeIP::Message const&) const (__args#0=...,
    this=0xaaaaaab11ed0)
    at /usr/include/c++/12.3.0/bits/std_function.h:591
#14 CommonAPI::SomeIP::Connection::handleStubReceive (this=0xaaaaaab11ce0,
    _message=...)
    at /home/user/capicxx-someip-runtime/src/CommonAPI/SomeIP/Connection.cpp:165
#15 0x0000fffff7633538 in std::function<void ()>::operator()() const (
    this=<optimized out>)
    at /usr/include/c++/12.3.0/bits/std_function.h:591
#16 vsomeip_v3::application_impl::invoke_handler (
    this=this@entry=0xaaaaaab12450, _handler=...)
    at /home/user/vsomeip/implementation/runtime/src/application_impl.cpp:2013
#17 0x0000fffff7633f48 in vsomeip_v3::application_impl::main_dispatch (
    this=0xaaaaaab12450)
    at /home/user/vsomeip/implementation/runtime/src/application_impl.cpp:1798
#18 0x0000fffff79e0d80 in ?? () from /usr/lib/libstdc++.so.6
#19 0x0000fffff779db34 in start_thread (arg=0xfffff5efe32f)
    at pthread_create.c:444
#20 0x0000fffff78058dc in thread_start ()
    at ../sysdeps/unix/sysv/linux/aarch64/clone.S:79
```

**Reply by srl100** [reply]
Anyone?

**Reply by srl100** [reply]
NB: This service is being accessed over UDP.

I have backtraced the call tree up as far as StubManager::handleMessage(const Message &_message) and even here I see that _message.getClientId() returns 0, but I've not really got the skills to trace this futher back up the stack!

**Reply by srl100** [reply]
Comments added after issue https://github.com/COVESA/vsomeip/issues/679 was closed appear to be relevant to this issue.

How are other people determining the peer address for a service accessed over UDP ?

**Reply by duartenfonseca** [reply]
Hi @srl100 , as it was mentioned here https://github.com/COVESA/vsomeip/issues/679 the method getHostAddress() would only work with internal TCP connection (applications on the same device). This does not seem to be what you are hoping to achieve. CommonAPI is supposed to work in a abstraction layer above network communication, in the sense that it does not have interfaces for dealing with it. 

> CommonAPI C++ is a C++ framework for interprocess and network communication. The basic objective is not to provide a new mechanism for interprocess or network communication (IPC), but to define a high-level C++ API, which can be used for different mechanisms. The benefit for C++ developers should be that they are relieved from details of proprietary or little-known communication frameworks or protocols. 

this quote is from the CommonAPI wiki (https://covesa.github.io/capicxx-core-tools/)

You would have to get the IP from another source

**Reply by srl100** [reply]
Hi @duartenfonseca , as it stands it seems to me that _client->getHostAddress() is fundamentally flawed - it does not give you the host address of the client - so we need a work around...

If I got rid of CommonAPI C++ and just used vsomeip on its own (which seems to be like a retrograde step) do you think that I would stand a chance at being able to determine the (IP) source of a request? What am I losing in taking this approach?

I can see that feature would be useful for several reasons (e.g. logging, rate limiting) aside from my purpose (implementing a temporary lock of a service to a particular client IP address).

I can see now that another issue I raised, not being able to set the SOME/IP response code, is also related to this abstracting away of the network communication - once we have got past the stub adapter checks of known services and methods and we are in the dispatcher/handler there appears to be no way to pass a response code back to the SOME/IP layer, so services such as VCIC, which use the SOME/IP response code to reject invalid parameters, cannot be implemented using CommonAPI.

**Reply by duartenfonseca** [reply, marked as answer]
- For CAPI the interface the _client->getHostAddress() will only return the routing HOST address of the local tcp communication. If unix domain sockets are used instead, there will be no address.

- vsomeip-lib also doesn't provide this as an interface.

The design of CommonAPI (and also vsomeip) is to keep away the user from the underlying network. That´s why all the IP addresses, ports, etc are provided by the vsomeip configuration and not by an API. If an application requires this information for whatever reason, these are the wrong libraries to use.
 
 The feature that could be added here on COVESA would be having the possibility to get that information for debugging purposes only
